File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        
        freq_ele = {}

        for n in nums:
            if n in freq_ele:
                freq_ele[n] += 1
            else:
                freq_ele[n] = 1

        bucket = [[] for _ in range(len(nums) + 1)]

        for num, freq in freq_ele.items():
            bucket[freq].append(num)

        result = []
        for i in range(len(bucket)-1, 0 , -1):
            for num in bucket[i]:
                result.append(num)
                if len(result) == k:
                    return result      

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.dart
class Solution {
  List<int> topKFrequent(List<int> nums, int k) {
      
      Map<int, int> freqMap = {};

      for(int num in nums) {
         freqMap[num] = (freqMap[num] ?? 0) + 1;
      }

      List<List<int>> bucket = List.generate(nums.length + 1, (_) => []);

      freqMap.forEach((num, freq) {
         bucket[freq].add(num);
      });

      List<int> result = [];

      for(int i = bucket.length - 1 ; i >= 0 ; i--) {
          for(int num in bucket[i]) {
            result.add(num);
            if(result.length == k) return result;
          }
      }
    return result;     
  }
}